akka-remote/src/main/scala/akka/remote/artery/SendQueue.scala

```diff
@@ -42,26 +43,37 @@ private[remote] final class SendQueue[T] extends GraphStageWithMaterializedValue
 
   override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, QueueValue[T]) = {
     @volatile var needWakeup = false
-    @volatile var queue: Queue[T] = null
+    @volatile var producerQueue: Queue[T] = null
+    val queuePromise = Promise[Queue[T]]()
     // TODO perhaps try similar with ManyToOneConcurrentLinkedQueue or AbstractNodeQueue
 
     val logic = new GraphStageLogic(shape) with OutHandler with WakeupSignal {
+
+      // using a local field for the consumer side of queue to avoid volatile access
+      private var consumerQueue: Queue[T] = null
+
       private val wakeupCallback = getAsyncCallback[Unit] { _ ⇒
         if (isAvailable(out))
           tryPush()
       }
 
       override def preStart(): Unit = {
-        needWakeup = true
+        implicit val ec = materializer.executionContext
+        queuePromise.future.foreach(getAsyncCallback[Queue[T]] { q ⇒
```

Comment:
Maybe accept a Try in the AsyncCallback? This will just silently leak if the Future fails (which is probably unlikely, but handling it is not hard).